infrastructure/src/main/java/org/corfudb/infrastructure/CompactorLeaderServices.java

```diff
@@ -0,0 +1,497 @@
+            CheckpointingStatus idleStatus = buildCheckpointStatus(StatusType.IDLE);
+
+            TransactionalContext.getCurrentContext().setPriorityLevel(CorfuMessage.PriorityLevel.HIGH);
+            txn.clear(checkpointingStatusTable);
+            txn.clear(activeCheckpointsTable);
+            //Populate CheckpointingStatusTable
+            for (TableName table : tableNames) {
+                txn.putRecord(checkpointingStatusTable, table, idleStatus, null);
+            }
+
+            // Also record the minToken as the earliest token BEFORE checkpointing is initiated
+            // This is the safest point to trim at since all data up to this point will surely
+            // be included in the upcoming checkpoint cycle
+            long minAddressBeforeCycleStarts = corfuRuntime.getAddressSpaceView().getLogTail();
+            txn.putRecord(checkpointTable, DistributedCompactor.CHECKPOINT_KEY,
+                    RpcCommon.TokenMsg.newBuilder()
+                            .setEpoch(this.epoch)
+                            .setSequence(minAddressBeforeCycleStarts)
+                            .build(),
+                    null);
+
+            CheckpointingStatus newManagerStatus = buildCheckpointStatus(StatusType.STARTED, tableNames.size(),
+                    System.currentTimeMillis()); // put the current time when cycle starts
+            txn.putRecord(compactionManagerTable, DistributedCompactor.COMPACTION_MANAGER_KEY, newManagerStatus, null);
+
+            txn.commit();
+        } catch (Exception e) {
+            syslog.error("triggerDistributedCheckpoint hit an exception {}. Stack Trace {}", e, e.getStackTrace());
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Validates the liveness of any on-going checkpoint of a table
+     * ActiveCheckpointTable contains the list of tables for which checkpointing has started. This method is scheduled
+     * to execute continuously by the leader, which monitors the checkpoint activity of the tables present in
+     * ActiveCheckpointTable.
+     * if there are no tables present,
+     * ... check the CheckpointStatusTable for any progress made (this step is done to include the
+     * tables which are checkpointed really fast)
+     * ... if there's no progress for timeout ms, call finishCompactionCycle() to mark the end of the cycle
+     * if there are any slow checkpointers,
+     * ... monitor checkpointing of the table by observing if the checkpointStream's tail moves forward
+     * ... if it does not move forward for timeout ms, then mark it as failed
+     * Also, when checkpoint of a table is found to be failed, the cycle is immediately marks as failed.
+     *
+     * @param timeout in ms
+     */
+    public void validateLiveness(long timeout) {
+        List<TableName> tableNames;
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            tableNames = new ArrayList<>(txn.keySet(activeCheckpointsTable)
```

Comment:
if you already use Collectors.toList, why you need to create another copy of a list?

Reply:
Thanks for the catch.